File: P1/Visitante/visitante/VisitantePrecio.java
/** 
 *	Fichero VisitantePrecio.java, visitante, P1 -Patrones-
 *
 *
 *	06/03/2018
 *
 *	Universidad de Granada - Grado en Informatica
 *
 *	Asignatura: Desarrollo de Software
 *	Práctica 1: Ejercicio 3
 * 
 *	@author Alejandro Ruiz Becerra
 * 
 */
package visitante;

/**
 * @author Alejandro Ruiz Becerra
 *
 */
public class VisitantePrecio extends VisitanteEquipo{

	private double precio;
	private Cliente cliente;
	
	
	/**
	 * Constructor
	 * 
	 * @param c
	 */
	public VisitantePrecio(Cliente c) {
		precio = 0;
		cliente = c;
	}

	@Override
	public void visitarDisco(Disco d) {
		precio += d.precioConDescuento(cliente);
	}


	@Override
	public void visitarTarjeta(Tarjeta t) {
		precio += t.precioConDescuento(cliente);
	}


	@Override
	public void VisitarBus(Bus b) {
		precio += b.precioConDescuento(cliente);
	}
	
	public double precio() {
		return precio;
	}

}

File: P1/Sesion2/Visitante/src/VisitantePrecio.cpp
#include "VisitantePrecio.h"
#include "Bus.h"
#include "Disco.h"
#include "Tarjeta.h"

float VisitantePrecio::obtenerPrecioTotal() const {
    return total;
}

void VisitantePrecio::visitarBus(const Bus & bus) {
    total += bus.getPrecio();
}

void VisitantePrecio::visitarDisco(const Disco & disco) {
    total += disco.getPrecio();
}

void VisitantePrecio::visitarTarjeta(const Tarjeta & tarjeta) {
    total += tarjeta.getPrecio();
}